### PR TITLE
Add R6 southwest Oregon (SWO) Ecology dataset

### PR DESCRIPTION
**Pull request by grovduck**
This PR would add the R6 southwest Oregon (SWO) Ecology plot dataset to the sample data available in `sknnr`.  The dataset consists of a species matrix (percent cover by tree species) and an environmental matrix (climate, topography, and spectral) for 3,005 plots in SWO measured in 2000.  The R6 Ecology Program installed these plots and this PR is contingent upon their permission to use these data.  

This PR adds a new function (`load_swo_ecoplot`) which makes these data available to all `sknnr` (and `scikit-learn`) estimators.

~~Pending approval, the data in `swo_ecoplot_env.csv` and `swo_ecoplot_spp.csv` are all dummy values, but do currently pass our test suite.~~ EDIT: Actual data is now committed after approval from R6 Ecology Group and data citation is added.

### CODE REVIEW

File: src/sknnr/datasets/_base.py [hunk 1]
```diff
@@ -126,22 +173,55 @@ def load_moscow_stjoes(
...
+
+    Returns
+    -------
+    Dataset or tuple of ndarray
+        A Dataset object containing the data, target, and feature names. If return_X_y
+        is True, return a tuple of data and target arrays instead.
+
+    Notes
+    -----
+    TBD
```
Comment (grovduck) [inline src/sknnr/datasets/_base.py:213]: Obviously, I need to revisit these TBDs once we hear back.

File: tests/test_datasets.py [hunk 2]
```diff
@@ -1,74 +1,108 @@
...
-    assert y.shape == (165, 35)
+    assert X.shape == (configuration.n_samples, configuration.n_features)
+    assert y.shape == (configuration.n_samples, configuration.n_targets)
     assert_array_equal(X.index.values, data.index)
     assert_array_equal(y.index.values, data.index)
 
 
-def test_asframe_raises_without_pandas():
+@pytest.mark.parametrize("configuration", CONFIGURATIONS, ids=CONFIGURATION_IDS)
+def test_asframe_raises_without_pandas(configuration: DatasetConfiguration):
```
Comment (grovduck) [inline tests/test_datasets.py:93]: Perhaps not necessary to test both configurations on this one - I don't know why one would raise and the other wouldn't.  But I was doing blanket replacement and chose to do this one as well.

File: tests/test_datasets.py [hunk 3]
```diff
@@ -1,74 +1,108 @@
...
 import pytest
 from numpy.testing import assert_array_equal
 
-from sknnr.datasets import load_moscow_stjoes
+from sknnr.datasets import load_moscow_stjoes, load_swo_ecoplot
 
 
-def test_load_moscow_stjoes():
+@dataclass
+class DatasetConfiguration:
```
Comment (aazuspan) [inline tests/test_datasets.py:14]: Nice clean solution!

File: tests/test_datasets.py [hunk 4]
```diff
@@ -1,74 +1,108 @@
...
+CONFIGURATIONS = [
+    DatasetConfiguration(
+        load_function=load_moscow_stjoes, n_samples=165, n_targets=35, n_features=28
+    ),
+    DatasetConfiguration(
+        load_function=load_swo_ecoplot, n_samples=3005, n_targets=25, n_features=18
+    ),
+]
+
+CONFIGURATION_IDS = ["moscow_stjoes", "swo_ecoplot"]
```
Comment (aazuspan) [inline tests/test_datasets.py:30]: Can we combine these into a dict with the IDs as keys and configs as values, just to ensure they stay synced?

File: tests/test_datasets.py (hunk 2 above)
Comment (aazuspan) [reply tests/test_datasets.py:93]: Yeah, not necessary but it doesn't do any harm. I'm good with that choice.

File: src/sknnr/datasets/_base.py [hunk 5]
```diff
@@ -88,6 +88,53 @@ def _load_csv_data(
     return index, data, data_names
 
 
+def _load_dataset_from_filenames(
```
Comment (aazuspan) [inline src/sknnr/datasets/_base.py:91]: What do you think about `_load_dataset_from_csvs`, just to be a little more specific? I don't have a strong opinion here, so whatever you prefer works.

**Review by aazuspan — COMMENTED** [review]
This looks great! I think generalizing the data loading and testing was a good choice, especially if we end up adding other datasets down the line. Much cleaner than the sklearn approach.

My comments are all very minor.

File: src/sknnr/datasets/_base.py [hunk 6]
```diff
@@ -88,6 +88,53 @@ def _load_csv_data(
...
+    *,
+    data_filename: str,
+    target_filename: str,
+    return_X_y: bool = False,
+    as_frame: bool = False,
+) -> tuple[NDArray[np.float64], NDArray[np.float64]] | Dataset:
+    """Generic loader for datasets with separate data and target files.
```
[the remark this answers is not in the document]
Comment (grovduck) [reply src/sknnr/datasets/_base.py]: Good idea!  I copied/pasted the docstring from the Notes section of `_load_csv_data` and added a sentence to note that both datasets' IDs should match.  Please add more detail if you think it needs it.

[the remark this answers is not in the document]
Comment (grovduck) [reply src/sknnr/datasets/_base.py]: Yep, you're right.  I've edited this, but not sure it's quite right.  Suggestions definitely welcome.

File: tests/test_datasets.py (hunk 4 above)
Comment (grovduck) [reply tests/test_datasets.py]: Perfect.  I've taken a stab at this.

File: src/sknnr/datasets/_base.py (hunk 5 above)
Comment (grovduck) [reply src/sknnr/datasets/_base.py]: I tried to marry both our suggestions.  I went with `_load_dataset_from_csv_filenames`.  A bit long, but my brain wasn't reading `csvs` as the plural of `csv` - more like "cis-vis" :smile:.

Comment (aazuspan) [reply src/sknnr/datasets/_base.py]: That's fair, looks good to me!

**Comment by grovduck** [comment]
Just FYI, I ran these data through yaImpute and created a bunch of new train/test files.  Then I clumsily copied `test_port.py` into a `swo_ecoplot` specific version and all tests passed.  I didn't think it was worth adding all of these files just to verify the porting logic, but we may want to keep the yaImpute output around if we end up getting back to #42?

**Comment by aazuspan** [comment]
Definitely reassuring to hear that everything passes with a second dataset, but I agree that it's not needed in the test suite. For the connection with #42, are you thinking that we would verify accuracy against yaImpute one last time before we switch to regression testing and ditch the port tests? That sounds reasonable to me!

**Comment by grovduck** [comment]
> For the connection with #42, are you thinking that we would verify accuracy against yaImpute one last time before we switch to regression testing and ditch the port tests? That sounds reasonable to me!

I'm not sure what I was thinking, but I did just spend a bit of time playing around with `syrupy` and I think I understand the main concepts there.  I'll put some follow up thoughts in #42.

**Comment by grovduck** [comment]
@aazuspan.  I think this is ready to go, but thought I'd see if you had any issues with the docstring for `load_swo_ecoplot`.

**Comment by aazuspan** [comment]
LGTM!
